lib/phlex/component.rb

```diff
@@ -1,27 +1,112 @@
 # frozen_string_literal: true
+require "digest"
+require "byebug"
 
 module Phlex
   class Component
     include Node, Context
 
+    class << self
+      def cache_key
+        @cache_key ||= {
+          template: checksum,
+          subcomponents: subcomponents.map(&:checksum),
+          ancestors: dependant_ancestor_checksums
+        }
+      end
+
+      def checksum
+        @checksum ||= Digest::MD5.hexdigest(file)
+      end
+
+      def subcomponents
+        @subcomponents ||= direct_subcomponents.reduce(Set.new) do |set, component|
+          set << component
+          component.subcomponents.each { set << _1 }
+          set
+        end
+      end
+
+      private
+
+      def dependant_ancestor_checksums
+        dependant_ancestors.map do |ancestor|
+          Digest::MD5.hexdigest(
+            File.read(
+              Module.const_source_location(ancestor.name).first
+            )
+          )
+        end
+      end
+
+      def dependant_ancestors
+        ancestors.map
+          .reject { [self, Object, BasicObject, Kernel].include? _1 }
+          .reject { _1.name.start_with? "Phlex::" }
+      end
+
+      def direct_subcomponents
+        file.scan(/component\s([A-Z]+[A-Za-z:]*)/).flatten.map { const_get(_1) }
+      end
+
+      def file
+        File.read(Module.const_source_location(self.name).first)
+      end
+    end
+
     module Initializer
       def initialize(*args, parent: nil, assigns: [], **kwargs, &block)
         @_parent = parent
+        @_content = block
 
         super(*args, **kwargs)
 
         copy_assigns
+      end
+    end
 
-        template(&block)
+    module Cache
+      def initialize(*args, cache: false, **kwargs, &block)
+        @_cache = cache
+        super(*args, **kwargs, &block)
+      end
+
+      def call
+        return super unless @_cache
+        Rails.cache.fetch(cache_key) { super }
+      end
+
+      def cache_key
+        if @_content
+          content = Digest::MD5.hexdigest(
+            File.read(
+              @_content&.source_location.first
+            )
+          )
```

Comment:
We need to get the sub components that might have been called inside the content block too.